Experiment Prototype and Slice Refactor

RemingtonRohel opened the pull request:
Decoupling of experiment prototype and experiment slices, as well as using pydantic to significantly reduce the testing load. Goes along with branch `slice_refactor` of `borealis_experiments`. I would like to add some more tests to make it more exhaustive. This branch currently passes all experiment unit tests, and can run on the lab computer.

I also changed the experiment unittests a bit - no more .csv list, all tests are assumed to be in `borealis/src/borealis_experiments/` and `borealis/src/borealis_experiments/testing_archive/`, with all experiments in testing_archive assumed to raise exceptions. The exception thrown and the regex expected are returned from a class method of the ExperimentPrototype child class defined within each test file. This keeps the tests more closely tied to their expected behaviour, and removes the unwieldy .csv file.

RemingtonRohel commented:
Also to note: I made `rxonly` a required field if `tx_beam_order` not set - i.e. the user must both omit `tx_beam_order` and set `rxonly` for a receive-only slice to run. Also, I have removed `*_wave_tables` and all associated functionality around it from `sample_building.py` and `sequences.py`, as they were inoperable anyways. We can add more complicated waveforms in a more elegant way when necessary.

kevinkrieger reviewed (approved):
Seems good to me, just add some comments I think

RemingtonRohel commented:
I've made some changes and would like you guys to re-review before approving. 

* Changed up the experiment_prototype directory structure a bit
  - Removed some unused files
  - Put sample_building and decimation_scheme together into a new `experiment_utils` folder
  - Put the only list_test we actually used directly into experiment_slice
* Added logging where all those TODOs were
* Updated the docstrings and documentation where asked, and per the new directory structure.
* Made decimation_scheme default in ExperimentSlice - previously was requiring it to be specified always

RemingtonRohel commented:
Oh and one more change since I noticed from old TODOs:
I took a bunch of redundant methods from the scan classes and make it one smart method in ScanClassBase. The method checks the type of whichever object calls it to figure out how to combine nested slices within the class.

tjk584 commented on src/experiment_prototype/scan_classes/averaging_periods.py:
```diff
@@ -10,11 +10,23 @@
+import inspect
+from pathlib import Path
 
+# third party
```
```suggestion
```
I like the structure of splitting imports into three separate categories. These repeated comments aren't really needed though, IMO.

tjk584 reviewed (changes requested):
Docs failed to build for this branch - error when building on ReadTheDocs.org is here: https://readthedocs.org/projects/borealis/builds/20121495/

After fixing that problem locally, a more complicated error showed up:

```console
updating environment: [new config] 61 added, 0 changed, 0 removed
WARNING: autodoc: failed to import module 'brian' from module 'src'; the following exception was raised:                                                                                                        
No module named 'build.debug.src'
WARNING: autodoc: failed to import module 'experiment_handler' from module 'src'; the following exception was raised:
Traceback (most recent call last):
  File "/home/radar/sphinx-env/lib64/python3.9/site-packages/sphinx/ext/autodoc/importer.py", line 58, in import_module
    return importlib.import_module(modname)
  File "/usr/lib64/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/radar/borealis/src/experiment_handler.py", line 28, in <module>
    from experiment_prototype.experiment_prototype import ExperimentPrototype
  File "/home/radar/borealis/src/experiment_prototype/experiment_prototype.py", line 27, in <module>
    from experiment_prototype.experiment_slice import ExperimentSlice, slice_key_set, hidden_key_set
  File "/home/radar/borealis/src/experiment_prototype/experiment_slice.py", line 101, in <module>
    freq_khz = confloat(ge=options.min_freq / 1e3, le=options.max_freq / 1e3)
TypeError: unsupported operand type(s) for /: 'min_freq' and 'float'


Exception occurred:
  File "/home/radar/borealis/src/experiment_prototype/experiment_utils/decimation_scheme.py", line 96, in __init__
    if len(self.stages) > options.max_filtering_stages:
TypeError: '>' not supported between instances of 'int' and 'max_filtering_stages'
The full traceback has been saved in /tmp/sphinx-err-00osb08c.log, if you want to report the issue to the developers.
Please also report this if it was a user error, so that a better error message can be provided next time.
A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
make: *** [Makefile:56: html] Error 2
```

To fix the first problem, just add `pydantic` to `docs/requirements.txt`. The second problem requires further investigation.

tjk584 reviewed (approved):
Fixed docs problem, all good now.

kevinkrieger commented on src/experiment_prototype/scan_classes/scan_class_base.py:
```diff
@@ -137,77 +151,74 @@ def slice_combos_sorter(list_of_combos, all_keys):
+                                              f"{bad_slices}")
+            disjoint_sets[i] = sorted(list(disjoint_sets[i]))   # Convert to a list
+
+        return disjoint_sets
```
algorithm checks out :+1:

kevinkrieger commented on src/experiment_prototype/scan_classes/scan_class_base.py:
```diff
@@ -137,77 +151,74 @@ def slice_combos_sorter(list_of_combos, all_keys):
 
-        list_of_combos = sorted(list_of_combos)
-        return list_of_combos
+        for k, interface_value in self.interface.items():
```
```suggestion
        for slice_ids_combo, interface_value in self.interface.items():
```

kevinkrieger replied:
Could use a var name that describes what the key actually is

RemingtonRohel replied:
Good idea

kevinkrieger commented on src/experiment_prototype/experiment_slice.py:
```diff
@@ -0,0 +1,665 @@
+from pydantic import (
+    validator, root_validator, conlist, conint, confloat, StrictBool, StrictInt, PositiveFloat
+)
+from scipy.constants import speed_of_light
```
You're on another dimension

kevinkrieger commented on src/experiment_prototype/experiment_prototype.py:
```diff
@@ -953,40 +956,13 @@ def build_scans(self):
+        log.info(f"Number of Sequences in Scan #1, Averaging Period #1: "
+                 f"{len(self.__scan_objects[0].aveperiods[0].sequences)}")
+        log.info("Number of Pulse Types in Scan #1, Averaging Period #1, Sequence #1: "
+                 "{len(self.__scan_objects[0].aveperiods[0].sequences[0].slice_dict)}")
```
```suggestion
                 f"{len(self.__scan_objects[0].aveperiods[0].sequences[0].slice_dict)}")
```

kevinkrieger reviewed (approved):
A couple typos I think, otherwise looks great
Less code, much simple.
